python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+
+Note that Matlab objects, enumerations and sparse data types are currently not converted and will give an "`Unrecognised input`" error.
+Likewise, Python objects or other data types not listed above returned by the Python functions will give an "`Unknown return type`" error.
+In principle both these objects can be wrapped using the mechanisms described above for Matlab objects (through a `MatlabProxyObject`)
+and below (through a `pythonclasswrapper`) for Python objects but this has not been implemented.
+
+## The SpinW-Brille interface
+
+Whilst the above mechanism works well for `multifit` to call a user defined Python model function or Euphonic,
+for the SpinW-Brille interface an extra level of wrapping is needed.
+This is because (after [ADR #9](../../documentation/adr/0009-brille-integration.md)) the Matlab SpinW code calls the Python Brille code directly.
+Furthermore SpinW requires access to the Brille (Python) `BZGrid*` objects, which is possible with the built-in Matlab-Python (`py.*`) interface
+but not allowed when using `call_python` because `mex` files can only return Matlab types.
+
+Therefore for the PACE Python interface, a wrapping mechanism and an architectural design is needed.
+First, as described in [ADR #12](../../documentation/adr/0012-matlab-python-wrapper.md), calls to Python from Matlab need to be abstracted into
+a separate library which would have different implementations for the Matlab interface (using `py.*`) and for `pyHorace` (without `py.*`).
+This means that the Matlab code to be compiled does not include any reference to `py.*` which would otherwise cause a crash in `pyHorace`.
+The `py.*` namespace, nonetheless, can be used in the interface library for the Matlab interface, and an 
+[implementation](https://github.com/mducle/horace-euphonic-interface/blob/light_wrapper/%2Beuphonic/light_python_wrapper.m)
+suited to the needs of Euphonic and Brillem is currently in review.
+
+The `pyHorace` implementation, on the other hand will rely on `call_python` with Matlab-side and Python-side wrappers.
+On the Python side, a [`WrappedPythonClass`](https://github.com/mducle/hugo/blob/master/pyHorace/FunctionWrapper.py#L27)
+is used to wrap any Python class such that its objects are stored in a module-global dictionary and only a string reference is passed to Matlab.
+Additional functions, `get_obj_prop` and `call_obj_method` as well as wrapped constructors are defined
+and stored in a module-global dictionary for access by `call_python`.
+On the Matlab side, a [`pyclasswrapper`](https://github.com/mducle/hugo/blob/master/src/pyclasswrapper.m) 
+is used to wrap the these constructors and property access functions so as to present the same interface to the Matlab code as the pure Matlab case.
+The point of these indirections is that only string references rather than objects are passed between Matlab and Python
+but the behaviour is similar to creating objects and accessing their properties and methods directly.
```

Comment:
This section is a little confusing, it says the above mechanism works for user Python functions and Euphonic and that a wrapper is needed for the SpinW-Brille case, but then the implementation example given is for the horace-euphonic-interface.

It might be worth explaining why this is needed for the euphonic-horace-interface too (and if this implementation example is going to be expanded for use with SpinW-Brille)

Reply:
Sorry I was eliding to different things here together.

One is the mechanism for Matlab code to call Python when the user is using Matlab (the "implementation" linked - which points to the library I want both `horace-euphonic-interface` and `brillem` to use).

The second is the mechanism for compiled Matlab code to call Python internally when the user is using Python. This is the one that applies specifically to the SpinW-Brille interface.

I clarify this.